port-chooser/src/main/java/org/terracotta/utilities/test/net/PortManager.java

```diff
@@ -0,0 +1,567 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terracotta.utilities.test.net;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.WeakReference;
+import java.net.BindException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketTimeoutException;
+import java.net.UnknownHostException;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.IntUnaryOperator;
+
+/**
+ * Manages TCP port reservation/de-reservation while attempting to avoid
+ * intra- and extra-JVM interference.
+ * <p>
+ * The port manager maintains two levels of reservation -- intra-JVM using
+ * an internal {@link BitSet} and inter-JVM using locks against a
+ * {@link java.io.RandomAccessFile}
+ * allocated in a system-wide, world-writable, system-dependent location.
+ * To avoid reservation/use races,
+ * <a href="https://en.wikipedia.org/wiki/Ephemeral_port">ephemeral ports</a>
+ * are not used.
+ * <p>
+ * All users of {@code PortManager} in a JVM must use the same {@code PortManager}
+ * instance to ensure proper intra-JVM reservations.
+ */
+public class PortManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PortManager.class);
+
+  private static final PortManager INSTANCE = new PortManager();
+
+  /**
+   * Gets the singleton instance of {@code PortManager} to use in a JVM.
+   * @return the singleton {@code PortManager} instance
+   */
+  public static PortManager getInstance() {
+    return INSTANCE;
+  }
+
+  private static final int MAXIMUM_PORT_NUMBER = 65535;
+  private static final int MINIMUM_ASSIGNABLE_PORT_COUNT = 1024;
+
+  private static final InetAddress LOCALHOST;
+  static {
+    InetAddress localHost = null;
+    try {
+      // Tests use "localhost" as the server host name which _does_ go through name resolution ...
+      localHost = InetAddress.getByName("localhost");
+    } catch (UnknownHostException e) {
+      LOGGER.warn("Unable to obtain an InetAddress for localhost via InetAddress.getByName(\"localhost\")", e);
+    }
+    LOCALHOST = localHost;
+  }
+
+
+  private final Random rnd = new SecureRandom();
+
+  private final BitSet reservedPorts = new BitSet();
+
+  /**
+   * Port allocation map.  A zero bit represents an available port.
+   */
+  private final BitSet portMap = new BitSet(MAXIMUM_PORT_NUMBER + 1);
```

Comment:
I think some of this can be cleaned up if the sense of this bitset is inverted.

Reply:
Looked into this -- seems that it's operationally quite expensive since `BitSet.get(int, int)` doesn't take a slice but makes a copy ...